source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -82,71 +96,83 @@ void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) const {
-  if (!token_->isIssuerSpecified(jwt_.iss_)) {
-    ENVOY_LOG(debug, "Jwt for issuer {} is not extracted from the specified locations", jwt_.iss_);
-    doneWithStatus(Status::JwtUnknownIssuer);
-    return;
-  }
+    // Check if token is extracted from the location specified by the issuer.
+    const bool has_issuer = issuer_opt_ ? jwt_.iss_ == issuer_opt_.value()
+                                        : tokens_.back()->isIssuerSpecified(jwt_.iss_);
+    if (!has_issuer) {
+      ENVOY_LOG(debug, "Jwt issuer {} does not match required", jwt_.iss_);
+      status = Status::JwtUnknownIssuer;
+      tokens_.pop_back();
+      continue;
+    }
 
-  // Check "exp" claim.
-  const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
-                                  std::chrono::system_clock::now().time_since_epoch())
-                                  .count();
-  // NOTE: Service account tokens generally don't have an expiration time (due to being long lived)
-  // and defaulted to 0 by google::jwt_verify library but are still valid.
-  if (jwt_.exp_ > 0 && jwt_.exp_ < unix_timestamp) {
-    doneWithStatus(Status::JwtExpired);
-    return;
-  }
+    // Check "exp" claim.
+    const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
+                                    std::chrono::system_clock::now().time_since_epoch())
+                                    .count();
+    // NOTE: Service account tokens generally don't have an expiration time (due to being long
+    // lived) and defaulted to 0 by google::jwt_verify library but are still valid.
+    if (jwt_.exp_ > 0 && jwt_.exp_ < unix_timestamp) {
+      status = Status::JwtExpired;
+      tokens_.pop_back();
+      continue;
+    }
 
-  // Check the issuer is configured or not.
-  jwks_data_ = config_->getCache().getJwksCache().findByIssuer(jwt_.iss_);
-  // isIssuerSpecified() check already make sure the issuer is in the cache.
-  ASSERT(jwks_data_ != nullptr);
+    // Check the issuer is configured or not.
+    jwks_data_ = config_->getCache().getJwksCache().findByIssuer(jwt_.iss_);
+    // isIssuerSpecified() check already make sure the issuer is in the cache.
+    ASSERT(jwks_data_ != nullptr);
 
-  // Check if audience is allowed
-  if (!jwks_data_->areAudiencesAllowed(jwt_.audiences_)) {
-    doneWithStatus(Status::JwtAudienceNotAllowed);
-    return;
-  }
+    // Check if audience is allowed
+    bool allowed = audiences_ ? audiences_->areAudiencesAllowed(jwt_.audiences_)
+                              : jwks_data_->areAudiencesAllowed(jwt_.audiences_);
 
-  if (jwks_data_->getJwksObj() != nullptr && !jwks_data_->isExpired()) {
-    verifyKey();
+    if (!allowed) {
+      status = Status::JwtAudienceNotAllowed;
+      tokens_.pop_back();
+      continue;
+    }
+
+    if (jwks_data_->getJwksObj() != nullptr && !jwks_data_->isExpired()) {
+      verifyKey();
+      return;
```

Comment:
When should we "return" and when should we "continue"?

If we returns,  should we continue to check the next token?

Reply:
I guess there could be some optimization here for allow failed as that is the only case where different issuers' remote key can be fetched at the same time.

Reply:
I have trouble with understand this loop. So if the token is invalid json string, does not match issuer_, does not match audiences or is expired, then the token is ignored, otherwise, it will be `verifyKey` which break the loop. Should (at least) mis-match audiences or expiration also yield authentication fail?

Also,

Reply:
The way this filter works is if just one tokens satisfies the JWT requirement the request is allowed. Unless it's allow_missing_or_failed, then all request are allowed. We assume that other tokens on the request is meant for some other upstream service. Correct me if I'm wrong here @qiwzhang.

This loop is trying to find just one JWT with valid claims to verify the signature.

Reply:
To be clear, assume the input has 3 tokens: A with different issuer, B is expired, and the C has invalid signature (verifyKey will fail). While the end result for `allow_missing_or_failed=false`, is that none of these is accepted (and authn will return non OK status), if the order of token is ABC, all will be checked, but if the order is CAB or ACB, then the loop will end after C. To me, B is considering as bad as C.

IMO, a more consistent implementation is removing the `while` loop in this function. The function will validate one token at a time, and use doneWithStatus to iterate to the next token.